### DISCUSSION TITLE
Legacy browser support(without Proxy)

### DISCUSSION THREAD

**Discussion by bloody-ux** [solidjs/solid#823] [Q&A]
After looking into the source code of Solid, we found `mergeProps` and `splitProps` need `Proxy` support.

Is it possible to use Solid with browsers don't support `Proxy`? Because we have users using Safari Mobile 9 on iOS or Chrome < 49 on Android.

I know there's [proxy-polyfill](https://github.com/GoogleChrome/proxy-polyfill) provided by chrome team. But it's lack some of features which Solid depends on.

**Comment by ryansolid** [comment]
`mergeProps` is used by spread in JSX. The others are based on your own usage. Non-Proxy versions wouldn't fully work. The properties on the source objects couldn't change and for `splitProps` the rest object wouldn't work.

I never built Solid in a non-proxy world to be fair. In 2016 when I started I figured by the time anyone used it browsers like Internet Explorer would be dead etc.

EDIT: with 1.6 most cases can be done without proxies. See: https://github.com/solidjs/solid/releases/tag/v1.6.0

**Reply by perqa** [reply]
There is another huge market segment where SolidJS would be a perfect fit - **if it supported legacy browsers**. I'm thinking of embedded browsers, like in SmartTVs, set-top-boxes, gaming consoles, etc. 95% of the apps on those devices are web apps. Those browsers lag behind many years, and they can't be updated by users; maybe if the manufacturer releases a firmware update which contains a newer browser version, but that is unusual. Many of those devices are very slow, and would really benefit from a performant framework with a small footprint, like SolidJS.

**Comment by bloody-ux** [comment]
So there's no way to rescue those users who are still using old version browsers, unless the developers carefully write code and avoid using spread and rest operators.

**Comment by atk** [comment]
> So there's no way to rescue those users who are still using old version browsers, unless the developers carefully write code and avoid using spread and rest operators.

You can still use babel to transpile the end result into something that will work in older browsers. The performance will be worse, though, so it would make sense to run it only if necessary and use the modern version whenever possible.

**Reply by perqa** [reply]
> Due to the limitations of ES5, Proxies cannot be transpiled or polyfilled. See support in [various JavaScript engines](https://kangax.github.io/compat-table/es6/#test-Proxy).

Source: https://babeljs.io/docs/en/learn#proxies

**Comment by maxmilton** [comment]
As @atk mentioned, you'll likely need to use a transpiler like babel to generate JS bundles anyway, especially for legacy browsers. That will also inject polyfills or transpile out things like `queueMicrotask` too.

If you were determined to use Solid you could write custom ESLint rules to prevent using spread on JSX elements or stores + write tests to verify no `new Proxy` in the output JS bundles. Another idea is you might be able create an alternative JSX spread implementation without Proxy that's non-reactive.

Solid is great and very forward-thinking, but TBH it wouldn't be my first choice when targeting old browser engines.

**Reply by steve-taylor** [reply]
> Solid is great and very forward-thinking, but TBH it wouldn't be my first choice when targeting old browser engines.

There’s a lost opportunity here. Do we really need the performance boost over React in modern browser runtimes? Almost never, I would imagine. It’s definitely needed in embedded runtimes such as on smart TVs and game consoles (looking at you, PS4), but the problem is that `Proxy` doesn’t enjoy universal support there.

Does Svelte fare any better in terms of exotic browser runtime support? I’m not a fan of Svelte, though.

**Reply by maxmilton** [reply]
For modern browsers, beyond runtime speed, more optimal code is still desirable because it can reduce battery drain on mobile devices. Additionally, apps created with Solid have smaller JS bundle sizes which save bytes for users on metered networks. Are these such a big deal compared to React? It depends on your app's use case.

As for CTV (Connected TV) applications (smart TVs, set-top boxes, game consoles, etc.), it's a tricky space depending on how old the devices you need to support are and so what their browser versions are. Sometimes you need to support a particularly funky device with crazy API specs. It's very common to build out solutions to swap out parts of code for certain platforms.

Svelte is a viable option, but it has been years since I've last used so perhaps things have changed. You would need to do your own research looking at the output code after building a demo app.

Usually for CTV I've chosen Preact. Because the framework is so small it's relatively easy to compile down to ES3 or polyfill browser APIs for certain JS runtimes. Also, having a mostly React-compatible API makes it easier to find developers who can work on it right away (recruiting in tech is a big problem in and of itself!).

**Comment by ryansolid** [comment]
With Solid 1.6 we don't require proxies for most cases. This was a big part of the release to better support these devices. See release notes: https://github.com/solidjs/solid/releases/tag/v1.6.0
